featurebyte/api/user_defined_function.py

```diff
@@ -0,0 +1,319 @@
+from featurebyte.api.api_handler.user_defined_function import UserDefinedFunctionListHandler
+from featurebyte.api.api_object_util import (
+    ForeignKeyMapping,
+    iterate_api_object_using_paginated_routes,
+)
+from featurebyte.api.feature_store import FeatureStore
+from featurebyte.api.savable_api_object import DeletableApiObject, SavableApiObject
+from featurebyte.api.user_defined_function_injector import (
+    FunctionAccessor,
+    UserDefinedFunctionInjector,
+)
+from featurebyte.config import Configurations
+from featurebyte.enum import DBVarType
+from featurebyte.exception import DocumentCreationError, InvalidSettingsError
+from featurebyte.logging import get_logger
+from featurebyte.models.base import PydanticObjectId, get_active_catalog_id
+from featurebyte.models.user_defined_function import FunctionParameter, UserDefinedFunctionModel
+from featurebyte.schema.user_defined_function import UserDefinedFunctionCreate
+
+logger = get_logger(__name__)
+
+
+def _get_active_feature_store_id(catalog_id: Optional[ObjectId] = None) -> Optional[ObjectId]:
+    # get the active feature store id
+    if catalog_id is None:
+        catalog_id = get_active_catalog_id()
+
+    client = Configurations().get_client()
+    response = client.get(f"/catalog/{catalog_id}")
+    if response.status_code == HTTPStatus.OK:
+        default_feature_store_ids = response.json()["default_feature_store_ids"]
+        if default_feature_store_ids:
+            assert len(default_feature_store_ids) == 1, "Only one default feature store is allowed."
+            return ObjectId(default_feature_store_ids[0])
+    return None
+
+
+def _synchronize_user_defined_function(
+    func_accessor: FunctionAccessor, route: str, feature_store_id: Optional[ObjectId]
+) -> None:
+    # synchronize all user-defined functions to the function accessor
+    name_to_udf_dict: dict[str, UserDefinedFunctionModel] = {}
+    try:
+        if feature_store_id is None:
+            feature_store_id = _get_active_feature_store_id()
+            if feature_store_id is None:
+                # Cannot synchronize user-defined functions without specifying a feature store
+                return
+
+        for udf_dict in iterate_api_object_using_paginated_routes(
+            route, params={"feature_store_id": feature_store_id}
+        ):
+            udf = UserDefinedFunctionModel(**udf_dict)
+            assert udf.name is not None, "User-defined function name cannot be None."
+            if udf.name in name_to_udf_dict and udf.catalog_id is None:
+                # if udf is global but there is already a UDF with the same name (which must be local),
+                # skip the global UDF as it is overridden by the local one
+                continue
+            name_to_udf_dict[udf.name] = udf
+    except InvalidSettingsError:
+        # ignore invalid settings error due to fail to connect to the server
+        logger.info("Failed to synchronize user-defined functions.")
```

Comment:
Is it intentional that we don't raise an error that is visible to the user here?

Reply:
Yes. This error will be thrown when we don't set the user profile at the beginning.